src/View/RoleFromAuthenticationIdentityDelegator.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/**
+ * @see       https://github.com/laminas/laminas-navigation for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-navigation/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-navigation/blob/master/LICENSE.md New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Laminas\Navigation\View;
+
+use Laminas\Authentication\AuthenticationServiceInterface;
+use Laminas\View\Helper\Navigation\AbstractHelper;
+use Psr\Container\ContainerInterface;
+
+class RoleFromAuthenticationIdentityDelegator
+{
+    /** @var string */
+    private $authenticationServiceName;
+
+    /** @var string */
+    private $getRoleMethodName;
+
+    public function __construct(
+        string $authenticationServiceName = AuthenticationServiceInterface::class,
+        string $getRoleMethodName = 'getRole'
+    ) {
+        $this->authenticationServiceName = $authenticationServiceName;
+        $this->getRoleMethodName         = $getRoleMethodName;
+    }
+
+    public static function __set_state(array $state): self
+    {
+        return new self(
+            $state['authenticationServiceName'] ??
+            AuthenticationServiceInterface::class,
+            $state['getRoleMethodName'] ?? 'getRole',
+        );
+    }
+
+    public function __invoke(
+        ContainerInterface $container,
+        string $name,
+        callable $callback,
+        array $options = null
+    ) {
+        $helper = $callback();
+
+        if (! $helper instanceof AbstractHelper) {
+            return $helper;
+        }
+
+        if (! $container->has($this->authenticationServiceName)) {
+            return $helper;
+        }
+
+        $authenticationService = $container->get(
+            $this->authenticationServiceName
+        );
+        if (! $authenticationService instanceof
+              AuthenticationServiceInterface) {
+            return $helper;
+        }
+
+        if (! $authenticationService->hasIdentity()) {
+            return $helper;
+        }
+
+        $identity = $authenticationService->getIdentity();
+
+        $role = null;
+        if (is_object($identity)
+            && method_exists($identity, $this->getRoleMethodName)
+        ) {
+            /** @var mixed $role */
+            $role = $identity->{$this->getRoleMethodName}();
+        }
+
+        $helper->setRole($role);
```

Comment:
This is potentially broken, as it enforces temporal coupling between when the authentication storage has an identity (and role), and when the helper is built.

It's done in good faith, but IMO it is quite broken :thinking:

Reply:
Still, this is done in the view layer: potentially OK, but only works in traditional php-fpm based apps, not long-running apps.

Reply:
> This is potentially broken, as it enforces temporal coupling between when the authentication storage has an identity (and role), and when the helper is built.

Right, therefore the different checks before.

> …but only works in traditional php-fpm based apps, not long-running apps.

Correct, but this applies to view helpers in general.